### PR TITLE
[GlobalOpt] Add reshape propagation and transpose fusion into linalg results to transpose propagation

### PR DESCRIPTION
**Pull request by Max191**
This PR enables fusion of transposes into the DPS init of linalg ops during PropagateLinalgTranspose. This also enables fusion by expansion and collapsing to propagate reshapes up and down during transpose propagation. After this PR, transpose propagation works as follows:
 - First, canonicalize named contractions by fusing nearby transpose operations to create new named ops.
 - Propagate both reshapes and transposes up, and fuse transposes into the DPS inits of linalg ops.
 - Propagate both reshapes and transposes down, and fuse transposes into the inputs of linalg ops.
 - Compose adjacent transpose operations into a single transpose.

### CODE REVIEW

**Comment by hanhanW** [comment]
Why do we propagate reshape and transpose within a single pass? Can it be two passes? We are planning to enable transpose propagation by default, and most of issues in https://github.com/openxla/iree/pull/16059 are addressed. I'm afraid that this is adding more issues to block the enablement.

**Comment by benvanik** [comment]
(we're really going to want to rewrite these kind of things to use the Explorer - local propagation is fine today because we generate single functions from frontends and inline everything, but we won't be forever - being able to propagate across calls and global loads/stores is going to be important and we don't want two (or more) ways of doing that - fine for now I guess, but keep that in mind that any local propagation is a subset of what we need at the whole-program level)

**Comment by Max191** [comment]
> Why do we propagate reshape and transpose within a single pass? Can it be two passes? We are planning to enable transpose propagation by default, and most of issues in #16059 are addressed. I'm afraid that this is adding more issues to block the enablement.

There are some reshapes blocking transpose fusion with matmuls in SDXL. The idea is to propagate reshapes at the same time as propagating transposes to open up more transpose fusions. @qedawkins added the reshape propagation part, so he may have additional details, but I'm not sure if there is a nice way to move the reshape propagation to a separate pass, since we first propagate up and then down.

**Comment by qedawkins** [comment]
> Why do we propagate reshape and transpose within a single pass? Can it be two passes? We are planning to enable transpose propagation by default, and most of issues in #16059 are addressed. I'm afraid that this is adding more issues to block the enablement.

When we have
```
%0 = linalg.generic tensor<?xf32>
%1 = tensor.expand_shape [[0, 1]] to tensor<?x?xf32>
%2 = linalg.transpose %1 permutation = [1, 0]
```
we cannot propagate the transpose through the reshape because the transpose is acting on the dimensions introduced by the reshape, however it is possible to first bubble the reshape up through the producer generic and then fuse the transpose. There isn't really a way to avoid doing both at the same time.

W.r.t enabling the propagation pass by default, we can land that before landing the changes here.

> (we're really going to want to rewrite these kind of things to use the Explorer - local propagation is fine today because we generate single functions from frontends and inline everything, but we won't be forever - being able to propagate across calls and global loads/stores is going to be important and we don't want two (or more) ways of doing that - fine for now I guess, but keep that in mind that any local propagation is a subset of what we need at the whole-program level)

I see this as supplementary to the patterns added here. Full model transpose analysis is a lot more than what this pass is supposed to do, and has the danger of introducing cost models. This pass is just supposed to fold away/fuse a few transposes already present in the model, and one such pattern can be to use the explorer to look across the call graph/global users to fold away a transpose. We aren't expected to handle CFGs either at this level, so I don't see how the existing local propagation patterns here are related to what the explorer offers.

**Comment by qedawkins** [comment]
Removing benchmark flags because I realized they don't do anything to this PR.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp [hunk 1]
```diff
@@ -366,6 +508,13 @@ class FuseTransposeWithGenericConsumer
     // To do the fusion, we can simply apply the permutation of the transpose
     // to the results of the associated input's indexing map, and then forward
     // the input to the transpose to the consumer generic.
+    auto maybeGenericOp =
+        getGenericOpOrGeneralizeContraction(rewriter, linalgOp);
```
Comment (qedawkins) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:512]: I would revert the changes here that are generalizing contractions. This functionality is currently locked behind `aggressive-transpose-propagation` because not all backends are able to handle arbitrary contractions like this yet.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp [hunk 2]
```diff
@@ -113,7 +114,147 @@ static void specializeGenericTransposeOp(RewriterBase &rewriter,
...
+      return failure();
+    }
+    auto linalgOp = dyn_cast<linalg::LinalgOp>(result.getOwner());
+    if (!linalgOp) {
+      return failure();
+    }
+
+    int64_t resultIndex = result.getResultNumber();
+    auto maybeGenericOp =
+        getGenericOpOrGeneralizeContraction(rewriter, result.getOwner());
```
Comment (qedawkins) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:187]: Probably similar principle here, we should lock anything that generalizes a linalg named op behind the aggressive propagation flag.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp [hunk 3]
```diff
@@ -584,51 +832,119 @@ void PropagateLinalgTransposePass::runOnOperation() {
...
+        return false;
+      }
+
+      // Do not reshape producer linalg op if it has more than one user.
+      auto producerLinalgOp = dyn_cast<linalg::LinalgOp>(producer);
+      if (!producerLinalgOp || !producerLinalgOp->hasOneUse()) {
+        return false;
+      }
+      // Only reshape generic ops and contraction ops.
+      if (!linalg::isaContractionOpInterface(producerLinalgOp) &&
```
Comment (qedawkins) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:873]: lock contraction generalization behind aggressive prop here.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp [hunk 4]
```diff
@@ -584,51 +832,119 @@ void PropagateLinalgTransposePass::runOnOperation() {
...
+                                                       reshapePropagationFn);
+    sinkingPatterns.insert<SinkTransposeThroughExtractSlice>(context);
+    sinkingPatterns.insert<SinkTransposeThroughExpandShape>(context);
+    sinkingPatterns.insert<FuseTransposeWithLinalgOpConsumer>(context);
+    populateNamedOpSinkingPatterns(context, sinkingPatterns);
+    sinkingPatterns.add<PropagateTransposeThroughUnaryElementwiseInput>(
+        context, /*benefit=*/2);
+
     if (enableAggressivePropagation) {
       sinkingPatterns.insert<GeneralizeInputTransposedNamedOp>(context);
```
Comment (qedawkins) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:949]: You can use a similar trick for aggressive bubbling by having a pattern that looks at the consumer of the named op and generalizes it if there is a transpose.

**Review by qedawkins — CHANGES_REQUESTED** [review]
Looks good, although I wrote a decent chunk of this code so could help to have another rewiewer. Main comments are about moving some of the functionality added here behind the aggressive propagation flag because we are generalizing some named ops.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp [hunk 5]
```diff
@@ -113,7 +114,147 @@ static void specializeGenericTransposeOp(RewriterBase &rewriter,
...
+// linalg.generic {
+//   indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0, d2)>,
+//                    affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
+//   ins(%0 : tensor<2x7x5>) outs(%1 : tensor<7x2x5>)
+//
+// %2 = linalg.transpose ... permutation = [0, 2, 1] :
+//                           tensor<7x2x5> -> tensor<7x5x2>
+// Becomes
+//
+// linalg.generic {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:159]: Is it possible to keep the outs indexing map still identity. That can cause issues later on.... Also this isnt bubbling the transpose up. It is kind of absorbing it in.... Questioning if we really need this. We can probably achieve this as a nemd op change? Can you give the real example that is driving this?

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp [hunk 6]
```diff
@@ -584,51 +832,119 @@ void PropagateLinalgTransposePass::runOnOperation() {
...
+      if (!isa<tensor::ExpandShapeOp>(consumer) &&
+          !isa<tensor::CollapseShapeOp>(consumer)) {
+        return false;
+      }
+
+      return consumer->hasOneUse() &&
+             llvm::isa<linalg::TransposeOp>(*(consumer->user_begin()));
+    };
+    RewritePatternSet bubblingPatterns(context);
+    linalg::populateFoldReshapeOpsByExpansionPatterns(bubblingPatterns,
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:888]: If reshapes are blocking things then we should run these before doing transpose propagation. Also note that the reshape by expansion and reshape by collapse are converse of each other. If you put both together and use the same control function that seems problematic to me

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp (hunk 6 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:888]: The control function here is to "only do it if the reshape (collapse or expand) is a consumer of the generic." So basically this will never loop infinitely as it only attempts to move reshapes up the SSA chain (and vice versa for pushing reshapes down the SSA chain).

We might need to restrict this in cases where the generic has multiple results.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp (hunk 5 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:159]: The outs indexing map is kept as an identity. We do need a test for the actual case that motivated this change though. For context now, the motivating case is this:

```mlir
%3 = linalg.matmul ins(%0, %1 : tensor<8192x640xf16>, tensor<640x640xf16>)
       outs(%2 : tensor<8192x640xf16>) -> tensor<8192x640xf16>
%expanded = tensor.expand_shape %3 [[0, 1], [2, 3]]
       : tensor<8192x640xf16> into tensor<2x4096x10x64xf16>
%5 = linalg.transpose ins(%expanded : tensor<2x4096x10x64xf16>)
       outs(%4 : tensor<2x10x4096x64xf16>) permutation = [0, 2, 1, 3]
```
